cpp/mrc/src/public/node/edge_builder.cpp

```diff
@@ -31,68 +34,91 @@
-    VLOG(2) << "- (" << source.source_type_hash() << ", " << sink.sink_type_hash() << ")";
+    // Get the egress
+    auto egress = source.get_egress_obj();
 
-    if (EdgeAdapterRegistry::has_source_adapter(source.source_type()))
-    {
-        auto adapter = EdgeAdapterRegistry::find_source_adapter(source.source_type());
+    // Set to the sink
+    sink.set_egress_obj(egress);
+}
 
-        // Try and build the handle
-        auto handle = adapter(source, sink, sink.ingress_handle());
-        if (handle)
-        {
-            return handle;
-        }
+std::shared_ptr<IngressHandleObj> EdgeBuilder::do_adapt_ingress(const EdgeTypePair& target_type,
+                                                                std::shared_ptr<IngressHandleObj> ingress)
+{
+    // Short circuit if we are already there
+    if (target_type.full_type() == ingress->get_type().full_type())
+    {
+        return ingress;
     }
 
-    // Fallback -- probably fail
-    try
+    // Next check the static converters
+    if (mrc::node::EdgeRegistry::has_converter(target_type.full_type(), ingress->get_type().full_type()))
     {
-        auto fn_converter = mrc::node::EdgeRegistry::find_converter(source.source_type(), sink.sink_type());
-        return fn_converter(ingress_handle);
-    } catch (std::runtime_error e)
-    {
-        // Last attempt, check if types are the same and return ingress handle.
-        if (source.source_type() == sink.sink_type())
+        try
         {
-            return ingress_handle;
-        }
+            auto fn_converter =
+                mrc::node::EdgeRegistry::find_converter(target_type.full_type(), ingress->get_type().full_type());
 
-        throw e;
+            auto converted_edge = fn_converter(ingress->get_ingress());
+
+            return std::make_shared<IngressHandleObj>(converted_edge);
+        } catch (std::runtime_error e)
+        {
+            // Last attempt, check if types are the same and return ingress handle.
+            if (target_type.full_type() == ingress->get_type().full_type())
+            {
+                return ingress;
+            }
+
+            throw e;
+        }
     }
-}
 
-std::shared_ptr<channel::IngressHandle> EdgeBuilder::ingress_for_source_type(
-    std::type_index source_type,
-    mrc::node::SinkPropertiesBase& sink,
-    std::shared_ptr<channel::IngressHandle> ingress_handle)
-{
-    if (EdgeAdapterRegistry::has_sink_adapter(sink.sink_type()))
+    // Start dynamic lookup
+    VLOG(2) << "Looking for edge adapter: (" << type_name(target_type.full_type()) << ", "
+            << type_name(ingress->get_type().full_type()) << ")";
+    VLOG(2) << "- (" << target_type.full_type().hash_code() << ", " << ingress->get_type().full_type().hash_code()
+            << ")";
+
+    // Loop over the registered adaptors
+    const auto& adaptors = EdgeAdapterRegistry::registered_ingress_adapters;
+
+    for (const auto& adapt : adaptors)
```

Comment:
Mirror of my comment in EdgeRegistry: can we utilize a double map here instead of iterating through all possible adapters, or is there something that makes this not possible? Maybe even a multi-map for each source type, so we're not testing so many possibilities.

Reply:
Discussed in person. Resolving.